myia/opt/clean.py

```diff
@@ -18,12 +34,26 @@ def _reabs(self, a: AbstractDict):
     return (yield AbstractTuple)(self(x) for x in a.entries.values())
 
 
-def erase_class(root, manager):
-    """Remove the Class type from graphs.
+@overload  # noqa: F811
+def _reabs(self, a: AbstractUnion):
+    return (yield AbstractTaggedUnion)(
+        [type_to_tag(opt), self(opt)] for opt in a.options
+    )
+
+
+def simplify_types(root, manager):
+    """Simplify the set of types that can be found in the graph.
 
-    * Class[x: t, ...] => Tuple[t, ...]
-    * getattr(data, attr) => getitem(data, idx)
-    * make_record(cls, *args) => make_tuple(*args)
+    * Replace AbstractClass by AbstractTuple:
+      * Class[x: t, ...] => Tuple[t, ...]
+      * getattr(data, attr) => getitem(data, idx)
+      * make_record(cls, *args) => make_tuple(*args)
+
+    * Replace AbstractUnion by AbstractTaggedUnion:
+      * Union[a, b, c, ...] => TaggedUnion[1 => a, 2 => b, 3 => c, ...]
+      * hastype(x, type) => hastag(x, tag)
+      *                  => bool_or(hastag(x, tag1), hastag(x, tag2), ...)
+      * unsafe_static_cast(x, type) => casttag(x, tag)
```

Comment:
I forgot to document that, but it also replaces AbstractDict by Tuples.

Reply:
I'll add it